storage/cloudspanner/tree_storage.go

```diff
@@ -0,0 +1,523 @@
+			subtreeTbl,
+			[]string{colTreeID, colSubtreeID, colRevision, colSubtree},
+			[]interface{}{t.treeID, st.Prefix, t._writeRev, stBytes},
+		)
+		if err := stx.BufferWrite([]*spanner.Mutation{m}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *treeTX) flushSubtrees() error {
+	return t.cache.Flush(t.storeSubtrees)
+}
+
+// Commit attempts to apply all actions perfomed to the underlying Spanner
+// transaction.  If this call returns an error, any values READ via this
+// transaction MUST NOT be used.
+// On return from the call, this transaction will be in a closed state.
+func (t *treeTX) Commit() error {
+	t.mu.Lock()
+	defer func() {
+		t.stx = nil
+		t.mu.Unlock()
+	}()
+
+	if t.stx == nil {
+		return ErrTransactionClosed
+	}
+	switch stx := t.stx.(type) {
+	case *spanner.ReadOnlyTransaction:
+		glog.V(1).Infof("Closed readonly tx %p", stx)
+		stx.Close()
+		return nil
+	case *spanner.ReadWriteTransaction:
+		return t.flushSubtrees()
+	default:
+		return fmt.Errorf("internal error: unknown transaction type %T", stx)
+	}
+}
+
+// Rollback aborts any operations perfomed on the underlying Spanner
+// transaction.
+// On return from the call, this transaction will be in a closed state.
+func (t *treeTX) Rollback() error {
+	t.mu.Lock()
+	defer func() {
+		t.stx = nil
+		t.mu.Unlock()
+	}()
+
+	if t.stx == nil {
+		return ErrTransactionClosed
+	}
+	return nil
+}
+
+func (t *treeTX) Close() error {
+	if t.IsOpen() {
+		if err := t.Rollback(); err != nil && err != ErrTransactionClosed {
+			glog.Warningf("Rollback error on Close(): %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+// IsOpen returns true iff neither Commit nor Rollback have been called.
+// If this function returns false, further operations may not be attempted on
+// this transaction object.
+func (t *treeTX) IsOpen() bool {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.stx != nil
+}
+
+// ReadRevision returns the tree revision at which the current STH was stored.
+func (t *treeTX) ReadRevision() int64 {
+	sth, err := t.currentSTH(context.TODO())
+	if err != nil {
+		panic(err)
+	}
+	return sth.TreeRevision
+}
+
+// WriteRevision returns the tree revision at which any tree-modifying
+// operations will write.
+func (t *treeTX) WriteRevision() int64 {
+	rev, err := t.writeRev(context.TODO())
+	if err != nil {
+		panic(err)
+	}
+	return rev
+}
+
+// nodeIDToKey returns a []byte suitable for use as a primary key column for
+// the subtree which contains the id.
+// If id's prefix is not byte-aligned, an error will be returned.
+func subtreeKey(id storage.NodeID) ([]byte, error) {
+	// TODO(al): extend this check to ensure id is at a tree stratum boundary.
+	if id.PrefixLenBits%8 != 0 {
+		return nil, fmt.Errorf("id.PrefixLenBits (%d) is not a multiple of 8; it cannot be a subtree prefix", id.PrefixLenBits)
+	}
+	return id.Path[:id.PrefixLenBits/8], nil
+}
+
+// getSubtree retrieves the subtree specified by id at (or below) the requested revision.
```

Comment:
most recent at (or below)? to match comment for GetMerkleNodes()

Reply:
done